openSearchLane.go

```diff
@@ -0,0 +1,425 @@
+
+	osl2 := openSearchLane{
+		openSearchConnection: osl.openSearchConnection,
+	}
+	osl.openSearchConnection.attach()
+
+	ll.clone(&osl2.logLane)
+	osl2.logLane.writer = log.New(&osl2, "", 0)
+
+	return &osl2
+}
+
+func (osl *openSearchLane) DeriveWithCancel() (Lane, context.CancelFunc) {
+	childCtx, cancelFn := context.WithCancel(context.WithValue(osl.logLane.Context, parent_lane_id, osl.logLane.LaneId()))
+	ll := deriveLogLane(&osl.logLane, childCtx, osl.tees, osl.cr)
+
+	osl2 := openSearchLane{
+		openSearchConnection: osl.openSearchConnection,
+	}
+	osl.openSearchConnection.attach()
+
+	ll.clone(&osl2.logLane)
+	osl2.logLane.writer = log.New(&osl2, "", 0)
+
+	return &osl2, cancelFn
+}
+
+func (osl *openSearchLane) DeriveReplaceContext(ctx context.Context) Lane {
+	ll := deriveLogLane(&osl.logLane, context.WithValue(ctx, parent_lane_id, osl.LaneId()), osl.tees, osl.cr)
+
+	osl2 := openSearchLane{
+		openSearchConnection: osl.openSearchConnection,
+	}
+	osl.openSearchConnection.attach()
+
+	ll.clone(&osl2.logLane)
+	osl2.logLane.writer = log.New(&osl2, "", 0)
+
+	return &osl2
+}
+
+func (osl *openSearchLane) SetMetadata(key string, val string) {
+	osl.openSearchConnection.mu.Lock()
+	defer osl.openSearchConnection.mu.Unlock()
+
+	if osl.metadata == nil {
+		osl.metadata = make(map[string]string)
+		osl.metadata["timestamp"] = time.Now().UTC().Format(time.RFC3339)
+	}
+
+	osl.metadata[key] = val
+}
+
+func (osl *openSearchLane) Close() {
+	osl.openSearchConnection.detach()
+}
+
+func (osl *openSearchLane) Write(p []byte) (n int, err error) {
+	osl.openSearchConnection.mu.Lock()
+	defer osl.openSearchConnection.mu.Unlock()
+
+	logEntry := string(p)
+
+	if strings.Contains(logEntry, "\n") {
+		logEntry = strings.ReplaceAll(logEntry, "\n", "")
+	}
+
+	// TODO remove this line
+	fmt.Println(logEntry)
+
+	parentLaneId, ok := osl.Context.Value(parent_lane_id).(string)
+	if !ok {
+		parentLaneId = ""
+	}
+
+	logData := &OpenSearchLogMessage{
+		AppName:      osl.openSearchConnection.config.OpenSearchAppName,
+		ParentLaneId: parentLaneId,
+		JourneyID:    osl.journeyId,
+		LaneID:       osl.LaneId(),
+		LogMessage:   logEntry,
+		Metadata:     osl.metadata,
+	}
+
+	osl.openSearchConnection.logBuffer = append(osl.openSearchConnection.logBuffer, logData)
+
+	osl.openSearchConnection.messagesQueued++
+
+	if len(osl.openSearchConnection.logBuffer)%osl.openSearchConnection.config.LogThreshold == 0 {
+		osl.openSearchConnection.wakeCh <- struct{}{}
+	}
+
+	return len(p), nil
+}
+
+func (osl *openSearchLane) Stats() {
+	osl.openSearchConnection.mu.Lock()
+	defer osl.openSearchConnection.mu.Unlock()
+
+	logData := &OpenSearchLogMessage{
+		AppName:  "OpenSearchLane",
+		Metadata: make(map[string]string),
+	}
+	logData.Metadata["timestamp"] = time.Now().UTC().Format(time.RFC3339)
+	logData.Metadata["messagesSent"] = fmt.Sprint(osl.openSearchConnection.messagesSent)
```

Comment:
Stats() is not something to log. It needs to be given to the caller, and the caller decides what to do with it.

One of the more important purposes of statistics is ops monitoring. That is done with systems like Prometheus/Grafana. To implement that, the client pulls Stats(), and incorporates them into the rest of the service performance counters and gauges. That's all outside of Lane implementation.

Return a simple struct with the values here.